### PR TITLE
Job resource - controller logic

### PR DESCRIPTION
**Pull request by soltysh**
This is 3rd part of the job resource trilogy. Depends on #12910 and #13254.

@sdminonne @akram this is providing basic running mechanism for jobs, I'm still working on it, but it's getting better :wink: 
@smarterclayton @derekwaynecarr fyi

### CODE REVIEW

**Comment by soltysh** [comment]
Rebased against latest master to pick up the client changes, esp. the experimental part needed for testing.

**Comment by soltysh** [comment]
The logic itself, looks finished. All is left is integration tests for the controller.

**Comment by erictune** [comment]
@bgrant0607 told me that he likes how you are computing successes and failures by listing pods.  So that can stay.

**Comment by erictune** [comment]
@soltysh
To have a complete Job resource for our v1.1 release, I think we need:

1. this PR
1.  e2e tests
1. user docs
1. examples
1. reaper and scaler support in kubectl

Anything else you can think of?
Do you already have work in progress for all those items? If not, I'd be happy to pick up one of those items and work on it.   Maybe docs?

**Comment by soltysh** [comment]
I currently working on:
* e2e tests
* reaper and scaler support in kubectl

It would be great if you could write the user docs, I do have some examples, so I can create a separate PR with that.

**Comment by erictune** [comment]
Okay, will work on user docs.

**Comment by soltysh** [comment]
@erictune this is rebased controller part (I've also added new testclient action to simplify testing client, the other commit in this PR). I'm currently working on e2e tests, reaper and scaler and examples. I'll create separate PRs for those, to make the review easier, unless you want me to include them in here?

**Comment by soltysh** [comment]
Closing and reopening to trigger all the tests to run.

**Comment by soltysh** [comment]
Shippable is green, as well as travis, I'm assuming this is jenkins flake.

**Comment by smarterclayton** [comment]
ok to test

**Comment by soltysh** [comment]
I don't see any error other than `Build timed out (after 45 minutes).`, am I wrong?

**Comment by smarterclayton** [comment]
Are you sure something you added isn't hanging?  Probably need to run the e2e yourself

**Comment by soltysh** [comment]
Yeah, will do that right away.

**Comment by soltysh** [comment]
Rebased, I've run the e2e tests but after 4h(!) with no success, I'm thinking the problems with jenkins failures are somehow related to #13485.

**Comment by soltysh** [comment]
@erictune, so here's the current state:
* [x] reaper and scaler #14046 
* [x] example job #14101
* [ ] e2e - WIP

File: pkg/controller/job/job_controller.go [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+	successful = filterPods(pods, api.PodSucceeded)
+	if restartPolicy == api.RestartPolicyNever {
+		unsuccessful = filterPods(pods, api.PodFailed)
+	}
+	return
+}
+
+func (jm *JobManager) manageJob(activePods []*api.Pod, successful, unsuccessful int, job *experimental.Job) int {
+	active := len(activePods)
+	parallelism := *job.Spec.Parallelism
```
Comment (deads2k) [inline pkg/controller/job/job_controller.go]: Dereferencing without a nil check worries me.  Is everyone else satisfied with knowing that if the object is created via conversion from a serialized type and if no one ever clears the field, this works?  It worries me that this code depends on a particular path of building its objects.

**Comment by erictune** [comment]
do the e2e pass when you run them yourself, e.g. on vagrant?

File: pkg/controller/job/job_controller.go (hunk 1 above)
Comment (erictune) [inline pkg/controller/job/job_controller.go]: seems okay to me to skip nil checks in private methods.

File: pkg/controller/job/job_controller.go [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+			ListFunc: func() (runtime.Object, error) {
+				return jm.kubeClient.Experimental().Jobs(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return jm.kubeClient.Experimental().Jobs(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&experimental.Job{},
+		replicationcontroller.FullControllerResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
```
Comment (erictune) [inline pkg/controller/job/job_controller.go]: In a follow-up PR, consider factoring below inline function to a named function.

File: pkg/controller/job/job_controller.go [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+			AddFunc: jm.enqueueController,
+			UpdateFunc: func(old, cur interface{}) {
+				job := cur.(*experimental.Job)
+				for _, c := range job.Status.Conditions {
```
Comment (erictune) [inline pkg/controller/job/job_controller.go]: In a follow-up PR, consider a comment explaining why completed job changes are ignored.

File: pkg/controller/job/job_controller.go [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+				}
+			}(i)
+		}
+		wait.Wait()
+
+	} else if active < parallelism {
+		// how many executions are left to run
+		diff := *job.Spec.Completions - successful
+		// for RestartPolicyNever we need to count unsuccessful pods as well
+		if job.Spec.Template.Spec.RestartPolicy == api.RestartPolicyNever {
```
Comment (erictune) [inline pkg/controller/job/job_controller.go]: I think people will want a way to allow for a variable but finite number of failures, and a fixed number of successes.    So, I think we may want to revisit this logic before this graduates from experimental to v1 API.

**Comment by erictune** [comment]
LGTM.

**Comment by erictune** [comment]
the separation of commits is fine.

File: pkg/controller/job/job_controller.go [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+			diff -= unsuccessful
```
Comment (erictune) [inline pkg/controller/job/job_controller.go]: I think this differs from the API docs.   Fix the API docs in a separate PR?

File: pkg/controller/job/job_controller.go [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+
+// getPodJob returns the job managing the given pod.
+func (jm *JobManager) getPodJob(pod *api.Pod) *experimental.Job {
+	jobs, err := jm.jobStore.GetPodJobs(pod)
+	if err != nil {
+		glog.V(4).Infof("No jobs found for pod %v, job manager will avoid syncing", pod.Name)
+		return nil
+	}
+	// TODO: add sorting and rethink the overlaping controllers, internally and with RCs
+	return &jobs[0]
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: If there are more than one jobs for a pod, this is non-deterministic. This needs a sort by creation time stamp. It's a user/configuration error but this also needs a log statement.

File: pkg/controller/job/job_controller.go (hunk 5 above)
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: Or make this match the doc and the design, and we can revisit adding failed completion after further design discussion.

File: pkg/controller/job/job_controller.go [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+	"k8s.io/kubernetes/pkg/controller/replication"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type JobManager struct {
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: This should be called JobController

File: pkg/controller/job/job_controller.go [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+		wait.Add(diff)
+		for i := 0; i < diff; i++ {
+			go func(ix int) {
+				defer wait.Done()
+				if err := jm.podControl.DeletePod(job.Namespace, activePods[ix].Name); err != nil {
+					// Decrement the expected number of deletes because the informer won't observe this deletion
+					glog.V(2).Infof("Failed deletion, decrementing expectations for controller %q", jobKey)
+					jm.expectations.DeletionObserved(jobKey)
+					util.HandleError(err)
+					active++
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: This is not threadsafe.

File: pkg/controller/job/job_controller.go [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+		wait.Add(diff)
+		for i := 0; i < diff; i++ {
+			go func() {
+				defer wait.Done()
+				if err := jm.podControl.CreatePods(job.Namespace, job.Spec.Template, job); err != nil {
+					// Decrement the expected number of creates because the informer won't observe this pod
+					glog.V(2).Infof("Failed creation, decrementing expectations for controller %q", jobKey)
+					jm.expectations.CreationObserved(jobKey)
+					util.HandleError(err)
+					active--
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: Again, not threadsafe.

File: pkg/controller/job/job_controller.go [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+		wait := sync.WaitGroup{}
+		wait.Add(diff)
+		for i := 0; i < diff; i++ {
+			go func() {
+				defer wait.Done()
+				if err := jm.podControl.CreatePods(job.Namespace, job.Spec.Template, job); err != nil {
+					// Decrement the expected number of creates because the informer won't observe this pod
+					glog.V(2).Infof("Failed creation, decrementing expectations for controller %q", jobKey)
+					jm.expectations.CreationObserved(jobKey)
+					util.HandleError(err)
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: What is this supposed to do in tests and in production?

File: pkg/controller/job/job_controller.go [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+	jobKey, err := controller.KeyFunc(job)
+	if err != nil {
+		glog.Errorf("Couldn't get key for job %#v: %v", job, err)
+		return 0
+	}
+
+	if active > parallelism {
+		diff := active - parallelism
+		jm.expectations.ExpectDeletions(jobKey, diff)
+		glog.V(2).Infof("Too many pods running job %q, need %d, deleting %d", jobKey, parallelism, diff)
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: Debug log, should be 4 or higher

File: pkg/controller/job/job_controller.go [hunk 12]
```diff
@@ -0,0 +1,452 @@
...
+
+		active -= diff
+		wait := sync.WaitGroup{}
+		wait.Add(diff)
+		for i := 0; i < diff; i++ {
+			go func(ix int) {
+				defer wait.Done()
+				if err := jm.podControl.DeletePod(job.Namespace, activePods[ix].Name); err != nil {
+					// Decrement the expected number of deletes because the informer won't observe this deletion
+					glog.V(2).Infof("Failed deletion, decrementing expectations for controller %q", jobKey)
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: glog.Errorf

File: pkg/controller/job/job_controller.go [hunk 13]
```diff
@@ -0,0 +1,452 @@
...
+		if job.Spec.Template.Spec.RestartPolicy == api.RestartPolicyNever {
+			diff -= unsuccessful
+		}
+		// limit to parallelism and count active pods as well
+		if diff > parallelism {
+			diff = parallelism
+		}
+		diff -= active
+		jm.expectations.ExpectCreations(jobKey, diff)
+		glog.V(2).Infof("Too few pods running job %q, need %d, creating %d", jobKey, parallelism, diff)
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: Debug log, should be 4 or higher

File: pkg/controller/job/job_controller.go [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+	}()
+
+	obj, exists, err := jm.jobStore.Store.GetByKey(key)
+	if !exists {
+		glog.Infof("Job has been deleted %v", key)
+		jm.expectations.DeleteExpectations(key)
+		return nil
+	}
+	if err != nil {
+		glog.Infof("Unable to retrieve job %v from store: %v", key, err)
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: Error log

File: pkg/controller/job/job_controller.go [hunk 15]
```diff
@@ -0,0 +1,452 @@
...
+		Type:               experimental.JobComplete,
+		Status:             api.ConditionTrue,
+		LastProbeTime:      util.Now(),
+		LastTransitionTime: util.Now(),
+	}
+}
+
+func getStatus(jobKey string, restartPolicy api.RestartPolicy, pods []api.Pod) (successful, unsuccessful int) {
+	successful = filterPods(pods, api.PodSucceeded)
+	if restartPolicy == api.RestartPolicyNever {
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: Pods can be failed even if they have restart policy never e.g. node failure.

File: pkg/controller/job/job_controller.go [hunk 16]
```diff
@@ -0,0 +1,452 @@
...
+func newCondition() experimental.JobCondition {
+	return experimental.JobCondition{
+		Type:               experimental.JobComplete,
+		Status:             api.ConditionTrue,
+		LastProbeTime:      util.Now(),
+		LastTransitionTime: util.Now(),
+	}
+}
+
+func getStatus(jobKey string, restartPolicy api.RestartPolicy, pods []api.Pod) (successful, unsuccessful int) {
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: jobKey is unused, why is it an argument to this?

File: pkg/controller/job/job_controller.go [hunk 17]
```diff
@@ -0,0 +1,452 @@
...
+		active -= diff
+		wait := sync.WaitGroup{}
+		wait.Add(diff)
+		for i := 0; i < diff; i++ {
+			go func(ix int) {
+				defer wait.Done()
+				if err := jm.podControl.DeletePod(job.Namespace, activePods[ix].Name); err != nil {
+					// Decrement the expected number of deletes because the informer won't observe this deletion
+					glog.V(2).Infof("Failed deletion, decrementing expectations for controller %q", jobKey)
+					jm.expectations.DeletionObserved(jobKey)
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: Deletion observerd after we failed to delete?

File: pkg/controller/job/job_controller.go [hunk 18]
```diff
@@ -0,0 +1,452 @@
...
+		active += diff
+		wait := sync.WaitGroup{}
+		wait.Add(diff)
+		for i := 0; i < diff; i++ {
+			go func() {
+				defer wait.Done()
+				if err := jm.podControl.CreatePods(job.Namespace, job.Spec.Template, job); err != nil {
+					// Decrement the expected number of creates because the informer won't observe this pod
+					glog.V(2).Infof("Failed creation, decrementing expectations for controller %q", jobKey)
+					jm.expectations.CreationObserved(jobKey)
```
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: Creation observed after we failed to create?

File: pkg/controller/job/job_controller.go (hunk 9 above)
Comment (mikedanese) [inline pkg/controller/job/job_controller.go]: Why do we decrement active when we fail to delete a pod?

**Comment by soltysh** [comment]
@mikedanese I'll be addressing your and @erictune's comments next week in a followup PR.

**Pull request merged**